sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/plans/ConstraintPropagationSuite.scala

```diff
@@ -217,4 +218,20 @@ class ConstraintPropagationSuite extends SparkFunSuite {
         IsNotNull(resolveColumn(tr, "a")),
         IsNotNull(resolveColumn(tr, "b")))))
   }
+
+  test("infer constraints on cast") {
```

Comment:
We need more test cases here. Any bug in this PR could cause query return a wrong answer. We need to be very cautious.